inflect/__init__.py

```diff
@@ -3860,25 +3838,21 @@ def number_to_words(  # noqa: C901
         else:
             chunks = [num]
 
-        first: Union[int, str, bool] = 1
+        first: Optional[bool] = True
         loopstart = 0
 
         if chunks[0] == "":
-            first = 0
+            first = None
             if len(chunks) > 1:
                 loopstart = 1
 
         for i in range(loopstart, len(chunks)):
-            chunk = chunks[i]
-            # remove all non numeric \D
-            chunk = NON_DIGIT.sub("", chunk)
-            if chunk == "":
-                chunk = "0"
+            chunk = "".join(filter(str.isdigit, chunks[i])) or "0"
```

Comment:
If a `str.retain(chars)` method existed, then this would be equivalent to:

```
chunk = chunks[i].retain(string.digits) or "0"
```

(at the moment, I don't think this line is as readable as it could be; my hope is that after further refactoring, there'll be a way to improve that)